### PR TITLE
Fix to use HW conversion instructions for fp8 casting on gfx941 and gfx942

### PR DESCRIPTION
**Pull request by wenchenvincent**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

### CODE REVIEW

**Review by wangye805 — COMMENTED** [review]
Quick question: why this change is needed with the cast_transpose PR? Our dev CI is not failing without this change.

**Comment by wenchenvincent** [comment]
> Quick question: why this change is needed with the cast_transpose PR? Our dev CI is not failing without this change.

This is for performance. They found that the current fp8 header was using the SW conversion on gfx942 and thus wrote their own conversion function. We should fix this in the fp8 header and let them use the fp8 header.

**Comment by wangye805** [comment]
> > Quick question: why this change is needed with the cast_transpose PR? Our dev CI is not failing without this change.
> 
> This is for performance. They found that the current fp8 header was using the SW conversion on gfx942 and thus wrote their own conversion function. We should fix this in the fp8 header and let them use the fp8 header.

If you take a look at the most recent CI: http://ml-ci-internal.amd.com:8080/job/TransformerEngine/job/TransformerEngine/job/PR-89/7/pipeline-console/?selected-node=356

**Comment by wenchenvincent** [comment]
> > > Quick question: why this change is needed with the cast_transpose PR? Our dev CI is not failing without this change.
> > 
> > 
> > This is for performance. They found that the current fp8 header was using the SW conversion on gfx942 and thus wrote their own conversion function. We should fix this in the fp8 header and let them use the fp8 header.
> 
> If you take a look at the most recent CI: http://ml-ci-internal.amd.com:8080/job/TransformerEngine/job/TransformerEngine/job/PR-89/7/pipeline-console/?selected-node=356

What's the issue?

**Comment by wangye805** [comment]
> > > > Quick question: why this change is needed with the cast_transpose PR? Our dev CI is not failing without this change.
> > > 
> > > 
> > > This is for performance. They found that the current fp8 header was using the SW conversion on gfx942 and thus wrote their own conversion function. We should fix this in the fp8 header and let them use the fp8 header.
> > 
> > 
> > If you take a look at the most recent CI: http://ml-ci-internal.amd.com:8080/job/TransformerEngine/job/TransformerEngine/job/PR-89/7/pipeline-console/?selected-node=356
> 
> What's the issue?

Those act test suites are failing: 
image

**Comment by wenchenvincent** [comment]
Seems it failed for this PR as well. Let me dig into it.

File: transformer_engine/common/amd_detail/hip_f8_impl.h [hunk 1]
```diff
@@ -126,8 +126,14 @@ uint8_t cast_to_f8(T _x, bool stoch, uint32_t rng) {
      but after shift right by 4 bits, it would look like midpoint.
   */
 
-  if (exponent_diff>0)
-    mantissa >>= exponent_diff;
+  if (exponent_diff>0) {
+    //mantissa >>= exponent_diff;
+    // Use a loop to work around the compiler bug that produced the wrong result for the right shift
+    // TODO: Change back after the compiler bug is resolved.
+    int shift_amount = exponent_diff > 32 ? 32 : exponent_diff;
```
Comment (ipanfilo) [inline transformer_engine/common/amd_detail/hip_f8_impl.h]: Is this limit sufficient by itself, w/o loop?

Comment (wenchenvincent) [reply transformer_engine/common/amd_detail/hip_f8_impl.h]: I tried to make it w/o loop and it worked. Code updated.

**Review by ipanfilo — APPROVED** [review]

File: transformer_engine/common/amd_detail/hip_f8_impl.h [hunk 2]
```diff
@@ -126,8 +126,12 @@ uint8_t cast_to_f8(T _x, bool stoch, uint32_t rng) {
      but after shift right by 4 bits, it would look like midpoint.
   */
 
-  if (exponent_diff>0)
-    mantissa >>= exponent_diff;
+  if (exponent_diff>0) {
+    //mantissa >>= exponent_diff;
+    // Work around the compiler bug that produced the wrong result for the right shift
+    // TODO: Change back after the compiler bug is resolved.
+    mantissa = exponent_diff >= 32 ? 0 : (mantissa >> exponent_diff);
```
Comment (draganmladjenovic) [inline transformer_engine/common/amd_detail/hip_f8_impl.h]: Driveby comment. Not sure if there is a compiler bug here. Out of range shifts are UB.

Comment (wenchenvincent) [reply transformer_engine/common/amd_detail/hip_f8_impl.h]: Thanks for the comment. Is there a reference?

Comment (ipanfilo) [reply transformer_engine/common/amd_detail/hip_f8_impl.h]: Indeed. C++ standard: 'If the value of rhs is negative or is not less than the number of bits in lhs, the behavior is undefined. '

Comment (wenchenvincent) [reply transformer_engine/common/amd_detail/hip_f8_impl.h]: Thanks. I updated the code and comments.

File: transformer_engine/common/amd_detail/hip_f8_impl.h [hunk 3]
```diff
@@ -118,16 +115,23 @@ uint8_t cast_to_f8(T _x, bool stoch, uint32_t rng) {
...
   /* This part is a bit tricky. The judgment of whether it is a tie needs to be done before we shift right 
      as shift right could rip off some residual part and make something not midpoint look like midpoint. 
      For example, the fp16 number 0x1002 (0 00100 0000000010), it is larger than midpoint, 
      but after shift right by 4 bits, it would look like midpoint.
   */
 
   if (exponent_diff>0)
-    mantissa >>= exponent_diff;
+    // Clip the exponent_diff as if the right shift when exponent_diff > 31 is undefined behavior
+    mantissa >>= (exponent_diff > 31 ? 31 : exponent_diff);
```
Comment (ipanfilo) [inline transformer_engine/common/amd_detail/hip_f8_impl.h:134]: It works because mantissa is actually shorter than 32 bits but strictly speaking there is no correct cutting of exponent_diff: 0x80000000 >> 31 = 1 while we would expect result of 0 for any shift value >31.

**Review by ipanfilo — APPROVED** [review]

File: transformer_engine/common/amd_detail/hip_f8_impl.h (hunk 3 above)
Comment (wenchenvincent) [reply transformer_engine/common/amd_detail/hip_f8_impl.h:134]: Yeah, you're right :)
